module3/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,241 @@
+        entityManager.persist(certificate);
+        return certificate;
+    }
+
+    private void saveTags(Set<Tag> tags) {
+        List<Tag> registerTags = tagDao.getAll();
+        getRegisterTagIds(tags, registerTags);
+        saveUnregisterTags(tags, registerTags);
+    }
+
+    private void getRegisterTagIds(Set<Tag> tags, List<Tag> registerTags) {
+        tags.stream()
+                .filter(registerTags::contains)
+                .forEach(tag -> tag.setId(registerTags.get(registerTags.indexOf(tag)).getId()));
+    }
+
+    private void saveUnregisterTags(Set<Tag> tags, List<Tag> registerTags) {
+        tags.stream()
+                .filter(tag -> !registerTags.contains(tag))
+                .forEach(tag -> tag.setId(tagDao.save(tag).getId()));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<GiftCertificate> getAll() {
+        return (List<GiftCertificate>) entityManager.createQuery("FROM GiftCertificate").getResultList();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<GiftCertificate> getAll(int startPosition, int recordsQuantity) {
+        return (List<GiftCertificate>) entityManager.createQuery("FROM GiftCertificate")
+                .setFirstResult(startPosition)
+                .setMaxResults(recordsQuantity)
+                .getResultList();
+    }
+
+    @Override
+    public GiftCertificate getById(int id) {
+        GiftCertificate certificate = entityManager.find(GiftCertificate.class, id);
+        entityManager.detach(certificate);
+        return certificate;
+    }
+
+    @Override
+    public GiftCertificate getByName(String name) {
+        return (GiftCertificate) entityManager.createQuery("FROM GiftCertificate c WHERE c.name = :name")
+                .setParameter("name", name)
+                .getSingleResult();
+    }
+
+    @Override
+    @SuppressWarnings("all")
+    public List<GiftCertificate> getByTagName(List<String> sortTypes, Set<String> names) {
+        String jpql = "SELECT c FROM GiftCertificate c JOIN c.tags t WHERE t.name ${conditions} ${value}";
+        Query query = entityManager
+                .createQuery(substituteJpqlQueryVariable(defineConditions(names), defineSortType(sortTypes), jpql));
+        setQueryParameters(query, names);
+        return query.getResultList();
+    }
+
+    private String defineConditions(Set<String> tagNames) {
+        return produceConditions(prepareConditions(tagNames));
+    }
+
+    private Set<String> prepareConditions(Set<String> conditions) {
+        return conditions.stream()
+                .filter(condition -> condition.trim().length() > 0)
+                .map(condition -> condition = ("=:").concat(condition.replaceAll(" ", "")))
+                .collect(Collectors.toSet());
+    }
+
+    private String produceConditions(Set<String> conditions) {
+        return String.join(CONDITIONS_TYPE.concat("t.").concat(SEARCHABLE_COLUMN), conditions);
```

Comment:
This code is confusing. It would be better to alias tags as tag for better readability.